dotnet-authserver/src/TeacherIdentity.AuthServer/Pages/SignIn/Register/CheckAccount.cshtml.cs

```diff
@@ -0,0 +1,122 @@
+using TeacherIdentity.AuthServer.Services.UserVerification;
+
+namespace TeacherIdentity.AuthServer.Pages.SignIn.Register;
+
+public class CheckAccount : PageModel
+{
+    private readonly IIdentityLinkGenerator _linkGenerator;
+    private readonly TeacherIdentityServerDbContext _dbContext;
+    private readonly IUserVerificationService _userVerificationService;
+    private readonly IClock _clock;
+
+    public CheckAccount(
+        IIdentityLinkGenerator linkGenerator,
+        TeacherIdentityServerDbContext dbContext,
+        IUserVerificationService userVerificationService, IClock clock)
+    {
+        _linkGenerator = linkGenerator;
+        _dbContext = dbContext;
+        _userVerificationService = userVerificationService;
+        _clock = clock;
+    }
+
+    public string? Email => HttpContext.GetAuthenticationState().EmailAddress;
+    public string? ExistingEmail => HttpContext.GetAuthenticationState().ExistingAccountEmail;
+
+    [BindProperty]
+    [Display(Name = " ")]
+    [Required(ErrorMessage = "Content TBD")]
+    public bool? IsUsersAccount { get; set; }
+
+    public void OnGet()
+    {
+    }
+
+    public async Task<IActionResult> OnPost()
+    {
+        if (!ModelState.IsValid)
+        {
+            return this.PageWithErrors();
+        }
+
+        var authenticationState = HttpContext.GetAuthenticationState();
+        authenticationState.OnExistingAccountConfirmed((bool)IsUsersAccount!);
+
+        if (IsUsersAccount == true)
+        {
+            await TryGenerateEmailPinForEmail(ExistingEmail!);
+            return Redirect(_linkGenerator.RegisterConfirmExistingAccount());
+        }
+
+        var user = await CreateUser();
+
+        authenticationState.OnUserRegistered(user);
+        await authenticationState.SignIn(HttpContext);
+
+        return Redirect(authenticationState.GetNextHopUrl(_linkGenerator));
+    }
+
+    private async Task<EmailValidationResult> TryGenerateEmailPinForEmail(string email)
+    {
+        var pinGenerationResult = await _userVerificationService.GenerateEmailPin(email);
+
+        switch (pinGenerationResult.FailedReasons)
+        {
+            case PinGenerationFailedReasons.None:
+                return EmailValidationResult.Success();
+
+            case PinGenerationFailedReasons.RateLimitExceeded:
+                return EmailValidationResult.Failed(new ViewResult()
+                {
+                    StatusCode = 429,
+                    ViewName = "TooManyRequests"
+                });
+
+            case PinGenerationFailedReasons.InvalidAddress:
+                ModelState.AddModelError(nameof(email), "Enter a valid email address");
+                return EmailValidationResult.Failed(this.PageWithErrors());
+
+            default:
+                throw new NotImplementedException($"Unknown {nameof(PinGenerationFailedReasons)}: '{pinGenerationResult.FailedReasons}'.");
+        }
+    }
+
+    private async Task<User> CreateUser()
```

Comment:
Considering creating a new Service for user creation

Reply:
We can loop back and look for repeated separately. The `TryGenerateEmailPinForEmail` above also is repeated in a few places. A series of helper classes `CreateUserHelper`, `EmailVerificationHelper` etc. may be the way to go.